src/engine/Union.h

```diff
@@ -85,4 +95,27 @@ class Union : public Operation {
   Result::Generator computeResultLazily(
       std::shared_ptr<const Result> result1,
       std::shared_ptr<const Result> result2) const;
+
+  // Compares two rows with respect to the columns that the result is sorted on.
+  template <size_t SPAN_SIZE>
+  bool isSmaller(const auto& row1, const auto& row2) const;
+
+  // Helper function for `computeResultKeepOrderImpl` that processes any
+  // remaining elements once one side is exhausted.
+  Result::Generator processRemaining(std::vector<ColumnIndex> permutation,
+                                     auto& it, auto end, bool requestLaziness,
+                                     size_t index, IdTable& resultTable,
+                                     LocalVocab& localVocab) const;
+
+  // Actual implementation of `computeResultKeepOrder`.
+  template <int COMPARATOR_WIDTH>
+  Result::Generator computeResultKeepOrderImpl(
+      bool requestLaziness, auto range1, auto range2,
+      std::pair<std::shared_ptr<const Result>, std::shared_ptr<const Result>>
+          lifetimeExtension) const;
+
+  // Similar to `computeResultLazily` but it keeps the order of the results.
```

Comment:
Is that the function that merges the results?
Then the name and the comment should reflect that.